policy: Add selector cache

jrajahalme opened the pull request:
Add an identity selector cache that does all the selector matching
necessary and caches the results.

This cache also gets updates from the kvstore on added and deleted
identities, and updates the selections accordingly.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7539)
<!-- Reviewable:end -->

jrajahalme commented on pkg/identity/cache/cache.go:
```diff
@@ -103,26 +102,64 @@ type identityWatcher struct {
+			deleted[id] = lbls
+		}
+		return true
+	} else {
```
All of the above fixed for the next push.

joestringer commented on pkg/identity/cache/cache.go:
```diff
@@ -103,26 +102,63 @@ type identityWatcher struct {
-					owner.TriggerPolicyUpdates(true, "one or more identities created or deleted")
-
+					if collectEvent(event, added, deleted) {
+						break
```
What does this `break` break?

joestringer replied:
(to answer my question, it's effectively a no-op; it breaks the switch which will exit immediately following this `collectEvent()` call anyway.)

joestringer replied:
I think this is intended to have a label like `waitForEvent:` above the select, and to `break waitForEvent`

jrajahalme replied:
Thanks for spotting this! Fixed and added unit tests to make sure all of this is covered and works as intended :-)

joestringer commented on pkg/identity/cache/cache.go:
```diff
@@ -103,26 +102,63 @@ type identityWatcher struct {
 			}
+
+			// see if there is more, but do not wait nor stop
+			select {
```
This only collects one more event. I suspect that you want something similar to the following incremental patch?

```
@@ -145,20 +145,23 @@ func (w *identityWatcher) watch(owner IdentityAllocatorOwner, events allocator.A
                        }
 
                        // see if there is more, but do not wait nor stop
-                       select {
-                       case event := <-events:
-                               // Collect more added and deleted labels
-                               switch event.Typ {
-                               case kvstore.EventTypeCreate, kvstore.EventTypeDelete:
-                                       collectEvent(event, added, deleted)
-                               case kvstore.EventTypeModify:
-                                       // Ignore modify events
+                       for {
+                               select {
+                               case event := <-events:
+                                       // Collect more added and deleted labels
+                                       switch event.Typ {
+                                       case kvstore.EventTypeCreate, kvstore.EventTypeDelete:
+                                               collectEvent(event, added, deleted)
+                                       case kvstore.EventTypeModify:
+                                               // Ignore modify events
+                                       }
+                               default:
                                }
-
-                       default:
-                               // Issue collected updates
-                               owner.UpdateIdentities(added, deleted)
                        }
+
+                       // Issue collected updates
+                       owner.UpdateIdentities(added, deleted)
+
                }
        }()
 }
```

(Seems like potentially a good target for a unit test to check that batching works correctly..)

jrajahalme replied:
Fixed!

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,335 @@
+}
+
+//
+// CachcedSelector implemenentation (== Public API)
```
```suggestion
// CachedSelector implementation (== Public API)
```

jrajahalme replied:
Fixed

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,335 @@
+}
+
+//
+// IdentitySelector implemenentation (== internal API)
```
```suggestion
// IdentitySelector implementation (== internal API)
```

jrajahalme replied:
Fixed

joestringer commented on pkg/identity/cache/cache.go:
```diff
@@ -103,26 +102,63 @@ type identityWatcher struct {
+		}
+		return true
+	}
+	log.Warningf("Ignoring unknown identity type '%s': %+v",
```
Please use the `log.WithFields(...).Warning(...)` style across this PR. For log management systems and even our CI, this makes errors easier to track. We now do things in the CI like counting the number of warnings with a unique message to figure out the most popular problems, so if the content of the log message has variable contents it can mess with that type of operation over the logs.

jrajahalme replied:
Done, mostly. For this one I did not, as this is something that should happen only if something is severely wrong with the source code, and could not readily figure out if there are suitable fields to use. Should I just invent new ones in this case?

joestringer replied:
(Sorry, missed this earlier) @jrajahalme if you think there's a case to reuse across multiple packages, be sure to add new fields to `pkg/logging/logfields/`, otherwise I'm fine with just defining the parameter titles using direct strings here.

Given the `%+v`, I think it's important to do so with this particular message.

joestringer reviewed (changes requested):
This is really slick! Core logic looks all good, I have just one note on event batching below. You might consider also putting a `Trigger` there too, just to put a bound on how often we cause churn in the subscribers.

I realize that the SelectorCache is written in such a way that concurrent updates will not cause problems (a mutex protects write + atomic read pointer to immutable state), but it'd be nice to see unit tests with concurrent writes and concurrent reads + writes.

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,335 @@
+// IdentitySelector is used as a map key, so it must not be implemented by a
+// map, slice, or a func, or a runtime panic will be triggered. In all
+// cases below IdentitySelector is being implemented by structs.
+type IdentitySelector interface {
```
By the way I think this can be hidden, `identitySelector` since the interface isn't used outside this package.

jrajahalme replied:
Will do for the next round.

jrajahalme replied:
Done!

ianvernon commented on pkg/labels/labels.go:
```diff
@@ -458,11 +458,17 @@ func (l Labels) ToSlice() []Label {
 
-// LabelArray returns the labels as label array
+// LabelArray returns the labels as label array, sorted by the key
 func (l Labels) LabelArray() LabelArray {
```
Add a unit test to ensure that the labels are correctly sorted.

jrajahalme replied:
Added in #7815

ianvernon commented on pkg/labels/array.go:
```diff
@@ -38,10 +42,19 @@ func ParseSelectLabelArray(labels ...string) LabelArray {
+}
+
+// NewLabelArrayFromSortedList returns labels based on the output of SortedList()
+func NewLabelArrayFromSortedList(list string) LabelArray {
```
Add a unit test here as well.

jrajahalme replied:
Added in #7815

ianvernon commented on pkg/labels/array.go:
```diff
@@ -38,10 +42,19 @@ func ParseSelectLabelArray(labels ...string) LabelArray {
-		array[i] = ParseLabel(base[i])
+	i := 0
+	for _, v := range base {
+		if lbl := ParseLabel(v); lbl.Key != "" {
```
Should we warn via a log message if the key is equal to an empty string? When would such a case happen?

jrajahalme replied:
The SortedList format terminates in a ';', which will produce an empty string when split. That's the only case I know of.

ianvernon commented on pkg/identity/cache/allocator.go:
```diff
@@ -33,22 +33,35 @@ import (
 // globalIdentity is the structure used to store an identity in the kvstore
 type globalIdentity struct {
-	labels.Labels
+	labels.LabelArray
```
All of these commits around labels-related changes seem to be better for another PR outside of this selector cache implementation. Can you factor this commit and the following into a separate PR?

* 7f000d77731aeab68f556826517239635a1af0d9
* 0ce43c1f569b8007050c56eff8b67b2718d05a38
* 90f4bea8b7a998658bf0350762fc0df5d19e9e74

jrajahalme replied:
@ianvernon Looking into this, but it seems that if we split this up at least one more of the remaining commits would need to be combined with these. This is due to the new function `identity.NewIdentityFromLabelArray()` introduced here only being used in the next commit. Will post a separate PR for the first 4 commits.

jrajahalme replied:
Done (#7815)

ianvernon commented on pkg/identity/cache/allocator.go:
```diff
@@ -33,22 +33,35 @@ import (
+		// We absolutely care that the final character is a semi-colon.
+		// Identity allocation in the kvstore depends on this (see
+		// kvstore.prefixMatchesKey())
+		str += fmt.Sprintf(`%s:%s=%s;`, l.Source, l.Key, l.Value)
```
@tgraf PTAL here - this looks good to me, but want to make sure you get sign-off on this as this is critical functionality

jrajahalme replied:
I factored this piece out so that this is only in one place, should be more straightforward to review!

ianvernon commented on pkg/identity/cache/allocation_test.go:
```diff
@@ -100,17 +104,120 @@ func (e *IdentityAllocatorConsulSuite) SetUpTest(c *C) {
+			if nid == target {
+				return rounds
+			}
+		case <-time.After(60 * time.Second):
```
This timeout would occur only for the given `select` statement (i.e., if nothing is sent to `d.updated`) vs. after an entire duration of 60 seconds for `target` to be reached for all iterations of receiving off of `d.updated` - is this intentional? Should the time be declared outside of the `for` loop, and then have `case <- (60 seconds since time outside of loop)`?

jrajahalme replied:
This is a testing utility, so the exact timeout semantics is not very significant. I kept the function as is and changed the function comment to end in:
```
// Returns 0 in case of 'd.updated' channel is closed or
// nothing is received from that channel in 60 seconds.
```

ianvernon commented on pkg/identity/cache/allocation_test.go:
```diff
@@ -212,6 +343,10 @@ func (ias *IdentityAllocatorSuite) TestLocalAllocationr(c *C) {
 
+	// KV-store still holds on to the key, so it is not deleted yet via KV-store events
+	// This may be racy, as timing here depends on scheduling.
+	c.Assert(owner.GetIdentity(id.ID), checker.DeepEquals, lbls1.LabelArray())
```
you can use `testutils.WaitUntil` here for functionality that is ran in the background but should return after some amount of time.

jrajahalme replied:
I don't know the exact timing, but the KV-store should not immediately release the key. This test is for testing that this is the case, not for the fact that the key is eventually deleted from the KV-store.

ianvernon commented on pkg/identity/cache/cache.go:
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
+			added[id] = gi.LabelArray
+			return true
+		}
+		log.Warningf("collectEvent: Ignoring unknown identity type '%s': %+v",
```
Use `log.WithFields` here and have fields for the type of the key / the key itself.

joestringer replied:
This one still doesn't appear to be resolved ^

ianvernon replied:
@jrajahalme ^^

ianvernon commented on pkg/identity/cache/cache.go:
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
 	stopChan chan bool
 }
 
+func collectEvent(event allocator.AllocatorEvent, added, deleted IdentityCache) bool {
```
Even though this function is not exported, I think this should be unit tested, as it is much easier to unit test than the  `watch` logic.

ianvernon commented on pkg/identity/cache/cache.go:
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
+			reflect.TypeOf(event.Key), event.Key)
+		return false
+	}
+	// Reverse an add when subsequently deleted
```
This code confused me, as it is implicit that we only reach this event if it is a `EventTypeDelete`, as the code always returns a value if the if-statement on line 109 evaluates to true. Can you make this a switch statement based off of the type of `event.Typ`? It makes the code clearer.

ianvernon replied:
@jrajahalme ^^

ianvernon commented on pkg/identity/cache/cache.go:
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
 	stopChan chan bool
 }
 
+func collectEvent(event allocator.AllocatorEvent, added, deleted IdentityCache) bool {
```
Add a comment indicating what the bool represents.

ianvernon replied:
@jrajahalme ^^

ianvernon commented on pkg/identity/cache/cache.go:
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
+					break More
+				}
 			}
+			// Issue collected updates
```
Should there be a `select` here if `w.stopChan` is closed? That way we don't have to issue identity updates to the owner.

ianvernon commented on pkg/identity/cache/cache.go:
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
-				return
+		More:
+			for {
+				// see if there is more, but do not wait nor stop
```
Can you elaborate on the rationale between having separate loops for `First` and `More` and why we care if `w.stopChan` is closed in `First` vs. not in `More`? Shouldn't we always stop if stopped?

jrajahalme replied:
The intent is to block or stop before the first event is received, and then pool as many other events together with the first one to amortize processing cost. The second loop does not block, and if we got an event, it was received before the stop channel was closed, so it should be processed. The stop channel will be tested again as soon as the first received event (together with any other events) is processed.

We could add check for the closure of the stop channel to the second loop, but IMO we still should process all events received before the stop channel was closed.

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+	// GetSelections returns the cached set of numeric identities
+	// selected by the CachedSelector.  The retuned slice must NOT
+	// be modified, as it is shared among multiple users.
+	GetSelections() []identity.NumericIdentity
```
Maybe wrap `[]identity.NumericIdentity` around a type and make the slice private? Then we can have an iterator function for it which returns the identities to those who need to iterate over it, and we can ensure that nothing can modify it without having to specify it here. This makes the modification impossible from outside users.

jrajahalme replied:
Let's see how we end up using this. Passing a slice onto which the numeric identities are appended by this function would be much simpler, if that satisfies the use case. But likely we'll be creating bpf map entries, one per numeric identity, and in that case iterator would be a good fit.

ianvernon replied:
👍 , we can always add this later.

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+// the CachedSelectors pushed by it.
+type CachedSelectionUser interface {
+	// IdentitySelectionUpdated implementations MUST NOT call back
+	// to selector cache while executing this function!
```
What is the reason for this requirement?

jrajahalme replied:
To avoid recursive/double locking. This function (callback) is called with the cache lock held, and calling any selector cache "API" function will panic.

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+}
+
+// NewSelectorCache creates a new SelectorCache.
+func NewSelectorCache() *SelectorCache {
```
This could be made private since it's not used anywhere outside of this package.

jrajahalme replied:
Right, will do.

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+	// selectorCache is the global selector cache. Additional ones
+	// are only created for testing.
+	selectorCache  = NewSelectorCache()
+	emptySelection = make([]identity.NumericIdentity, 0)
```
What does this global do? Please add a comment here.

jrajahalme replied:
Don't know how often this will be used, but is more cache efficient if we don't need multiple copies of empty slices (both slice header and the backing memory). We could maybe expand this to only keep one copy of any kind of a slice. We already keep them sorted, so it might be possible to build an index to them. I added this comment and changed the definition slightly as a nil slice should work as well:
```
	// Empty slice of numeric identities used for all selectors that select nothing
	emptySelection []identity.NumericIdentity
```

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+// selected.  The cached selections can be concurrently updated. In
+// that case GetSelections() will return either the old or new version
+// of the selections. If the old version is returned, the user is
+// guaranteed to receive a notification including the update.
```
including the update of the new version?

jrajahalme replied:
Right, there will be a notification of for the change that caused the new version to be created, but as updates are coalesced by the next time the user calls `GetSelections()` it may already be updated with further changes.

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+	selector         api.EndpointSelector
+	key              string
+	users            map[CachedSelectionUser]struct{}
+	selections       unsafe.Pointer
```
add a comment indicating that this will always be a slice of `identity.NumericIdentity`

jrajahalme replied:
done!

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+		selections[i] = nid
+		i++
+	}
+	// Sort the numeric identities so that the map iteration order
```
I don't understand this comment - map iteration order is random, so why do we need to sort?

jrajahalme replied:
Meaning that the order of the numeric identities in the slice will be the same regardless the what iteration order of the map happens to be. Makes testing and debugging easier.

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+// cached, the corresponding CachedSelector is returned, otherwise one
+// is created and added to the cache.
+func (sc *SelectorCache) AddIdentitySelector(user CachedSelectionUser, selector api.EndpointSelector) (cachedSelector CachedSelector, added bool) {
+	key := selector.LabelSelector.String()
```
Add a comment that that `String` may not always be identical for the case of requirements being in different order for `LabelSelector`.

jrajahalme replied:
Added this:
```
	// The key returned here may be different for equivalent
	// labelselectors, if the selector's requirements are stored
	// in different orders. When this happens we'll be tracking
	// essentially two copies of the same selector.
```

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+	// identities. Caller must use GetSelections() to get the
+	// current selections after adding a selector. This way the
+	// behavior is the same between the two cases here (selector
+	// is already cached, or is a new one).
```
We will have to account for this behavior for incremental updates, though. We will have to figure out how to communicate this downstream to identity policies in an elegant manner.

jrajahalme replied:
We may need to add a revision number to both the `GetSelections()` and incremental updates to properly sequence them on the user side.

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+		if _, exists := sc.idCache[numericID]; exists {
+			delete(sc.idCache, numericID)
+		} else {
+			log.WithFields(logrus.Fields{logfields.Identity: numericID}).Warningf("UpdateIdentities: Deleting a non-existing identity")
```
Warningf --> Warning

jrajahalme replied:
Thnx!

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+	}
+	for numericID, lbls := range added {
+		if _, exists := sc.idCache[numericID]; exists {
+			log.WithFields(logrus.Fields{logfields.Identity: numericID}).Warningf("UpdateIdentities: Adding an existing identity")
```
Warningf --> Warning

jrajahalme replied:
x2

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -0,0 +1,336 @@
+			}
+		case *fqdnSelector:
+			// TODO
+		}
```
Should we have a default case with an error? Not sure if this would ever happen, but could be helpful if someone adds another selector type in the future.

ianvernon replied:
@jrajahalme thoughts on this? ^ we may not even need this switch statement in the FQDN case since there is no logical way of corresponding a CIDR identity with a given FQDN by virtue of the fields of the identity itself at this current time - this relation has to come from the DNS proxy.

jrajahalme committed:
identity: Include event details also for local identities

Add the numeric ID and Key (labelarray) fields to the events generated
by the local identity allocations so that the watcher can know what
has happened. The actual labels are only needed for the create
event. The same globalIdentity type is used for this purpose as well.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Add selector cache

Add an identity selector cache that does all the selector matching
necessary and caches the results.

This cache also gets updates from the kvstore on added and deleted
identities, and updates the selections accordingly.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

ianvernon reviewed (approved):
Nothing blocking - this is solid enough for now that we can merge it, especially since this is not being plumbed yet throughout policy calculation. We will tune it as needed to meet our requirements. Ship it!
